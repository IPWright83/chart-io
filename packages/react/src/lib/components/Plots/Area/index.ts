export * from "./Area";
export * from "./StackedArea";
export * from "./Areas";
