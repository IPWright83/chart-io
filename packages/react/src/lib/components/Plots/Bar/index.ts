export * from "./Bar";
export * from "./StackedBars";
export * from "./GroupedBars";
export * from "./Bars";
