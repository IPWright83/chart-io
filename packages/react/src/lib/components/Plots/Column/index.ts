export * from "./Column";
export * from "./StackedColumn";
export * from "./GroupedColumn";
export * from "./Columns";
