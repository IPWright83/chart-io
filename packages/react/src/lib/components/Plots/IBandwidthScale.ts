export interface IBandwidthScale {
    bandwidth?: () => number;
}
