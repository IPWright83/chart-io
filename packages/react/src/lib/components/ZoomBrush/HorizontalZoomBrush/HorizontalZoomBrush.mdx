import { Story, Canvas, Meta } from "@storybook/blocks";

<Meta title="Components/Zoom/HorizontalZoomBrush" />

# HorizontalZoomBrush Component

The `<HorizontalZoomBrush>` component allows us to define a d3 Brush which allows us to zoom in on an area of the chart. This brush allows you to interact with it to both zoom and pan the main plot. It re-renders any plots in a minature fashion to provide context while zooming.

<Canvas>
    <Story id="components-zoom-horizontalzoombrush--default" />
</Canvas>
<br />

An example with a full chart can be seen here:

<Canvas>
    <Story id="xycharts-line--line-with-brush" />
</Canvas>

## Props

Generally you don't use the `<HorizontalZoomBrush>` component directly, instead prefering to use the `<ZoomBrush>` component with an `"inline"` type.

| Property   | Type          | Default | Note                                                    |
| ---------- | ------------- | ------- | ------------------------------------------------------- |
| `height`\  | `number`      | 60      | The height of the brush that will be reserved           |
| `children` | `JSX.Element` | []      | The plots that should be rendered in a minature version |
