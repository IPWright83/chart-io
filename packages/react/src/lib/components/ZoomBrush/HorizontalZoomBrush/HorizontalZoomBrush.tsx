import * as d3 from "@chart-it/d3";
import React, { useEffect, useRef } from "react";
import { useDispatch, useSelector } from "react-redux";

import { chartActions, chartSelectors, IState } from "../../../store";
import { logWarning } from "../../../utils";

export interface IHorizontalZoomBrushProps {
    /**
     * The child plots for the chart
     */
    plots?: JSX.Element[];

    /**
     * The height that the brush should be
     */
    height: number;
}

interface IScaleWithInvert {
    invert: any;
}

/**
 * Represents a Horizontal brush for zooming
 * @return The Horizontal Brush component
 */
export function HorizontalZoomBrush({ plots = [], height }: IHorizontalZoomBrushProps) {
    const x = plots.map((p) => p.props.x)[0];
    const ys = plots.flatMap((p) => p.props.y ?? p.props.ys);

    const brush = useRef();
    const dispatch = useDispatch();
    const xScale = useSelector((s: IState) => chartSelectors.scales.getScale(s, x, "brush"));
    const margin = useSelector((s: IState) => chartSelectors.dimensions.margin(s));
    const width = useSelector((s: IState) => chartSelectors.dimensions.width(s));
    const plotHeight = useSelector((s: IState) => chartSelectors.dimensions.height(s));

    // Update the range, to be used by any plots that are set to use a brush
    // This allows us to shrink and re-use the plots within the brush
    useEffect(() => {
        dispatch(chartActions.setBrushRange(x, [margin.left, width - margin.right]));
    }, [x, width, margin.right, margin.left]);

    // Reserve some space for the brush if it's visible
    useEffect(() => {
        const reservedHeight = plots.length > 0 ? height : 0;
        dispatch(chartActions.setBrushReservedDimensions(0, reservedHeight));
    }, [height, plots]);

    useEffect(() => {
        ys.forEach((key) => dispatch(chartActions.setBrushRange(key, [height, 0])));
    }, [ys]);

    useEffect(() => {
        // prettier-ignore
        if (brush.current) {
            const left = margin.left;
            const right = Math.max(left, width - margin.right);

            const xBrush = d3
                .brushX()
                .extent([[left, 0], [right, height]])
                .on("end", event => {
                    const extent = event.selection;
    
                    // Clear the brush 
                    if (!extent || !xScale) {
                        dispatch(chartActions.setScaleZoom(x, undefined));    
                        return;
                    }

                    // Ensure this scale works properly
                    if (!(xScale as IScaleWithInvert).invert) {
                        logWarning("W007", "An incompatible scale was used with a <HorizontalBrush>");
                        return;
                    }

// var eachBand = self.yScale.step();
// var index = Math.round((d3.event.y / eachBand));
// var val = self.yScale.domain()[index];

                    // Calculate the new domain and then zoom
                    const domain = [
                        (xScale as IScaleWithInvert).invert(extent[0]), 
                        (xScale as IScaleWithInvert).invert(extent[1])
                    ];
                    dispatch(chartActions.setScaleZoom(x, domain));
                });

            d3.select(brush.current).call(xBrush);
        }
    }, [dispatch, brush, width, height, margin.right, xScale]);

    // Clone each plot, as we need to re-render them within the brush
    // but ensure they don't trigger events, and they use the "brush" scale
    // to shrink them down to an appropriate size
    const plotClones = plots.map((plot, index) =>
        React.cloneElement(plot, { key: index, showInLegend: false, interactive: false, scaleMode: "brush" })
    );

    return (
        <>
            <g ref={brush} transform={`translate(0, ${plotHeight})`}>
                {plotClones}
            </g>
        </>
    );
}
