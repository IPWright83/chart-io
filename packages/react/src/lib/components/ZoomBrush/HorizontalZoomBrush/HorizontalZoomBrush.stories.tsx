import * as d3 from "@chart-io/d3";
import isChromatic from "chromatic/isChromatic";
import React from "react";
import { Provider } from "react-redux";

import { HorizontalZoomBrush } from ".";

import { createMockStorybookStore } from "../../../testUtils";
import { Line } from "../../Plots";

export default {
    title: "Components/Zoom/HorizontalZoomBrush",
    component: HorizontalZoomBrush,
    parameters: {
        chromatic: { delay: 300 },
    },
};

const HorizontalZoomBrushTemplate = (args) => {
    const store = createMockStorybookStore({
        chart: {
            dimensions: {
                width: 500,
                height: 0,
                plotMargin: { left: 0, top: 0, right: 0, bottom: 0 },
            },
            data: [
                { x: 10, y: 10 },
                { x: 50, y: 50 },
                { x: 250, y: 250 },
                { x: 450, y: 50 },
            ],
            scales: {
                x: {
                    scale: d3.scaleLinear(),
                    range: [0, 500],
                    domain: [0, 500],
                },
                y: {
                    scale: d3.scaleLinear(),
                    range: [0, 500],
                    domain: [0, 300],
                    brush: {
                        range: [0, 100],
                    },
                },
            },
        },
    });

    return (
        <Provider store={store}>
            <svg width="500px" height="100px" style={{ background: "white" }}>
                <HorizontalZoomBrush height={args.height}>
                    <Line x="x" y="y" />
                </HorizontalZoomBrush>
            </svg>
        </Provider>
    );
};

export const Default = {
    name: "Horizontal Zoom Brush",
    render: HorizontalZoomBrushTemplate,
    args: {
        height: 100,
    },
};
