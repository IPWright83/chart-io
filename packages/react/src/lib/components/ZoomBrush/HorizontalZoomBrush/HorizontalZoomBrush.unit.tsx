import * as d3 from "@chart-io/d3";
import React from "react";

import { renderChart, wait } from "../../../testUtils";
import { Line } from "../../Plots";

import { HorizontalZoomBrush } from "./HorizontalZoomBrush";

describe("HorizontalZoomBrush", () => {
    it("renders correctly", async () => {
        const data = [
            { x: 5, y: 5 },
            { x: 10, y: 10 },
        ];
        const scales = {
            x: d3.scaleLinear().domain([0, 20]).range([0, 100]),
            y: d3.scaleLinear().domain([0, 20]).range([0, 100]),
        };

        const { asFragment } = await renderChart({
            children: (
                <HorizontalZoomBrush height={100}>
                    <Line x="x" y="y" />
                </HorizontalZoomBrush>
            ),
            data,
            scales,
        });

        await wait(100);

        expect(asFragment()).toMatchSnapshot();
    });
});
