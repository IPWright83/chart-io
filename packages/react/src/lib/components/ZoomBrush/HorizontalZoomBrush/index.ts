import { HorizontalZoomBrush as HorizontalZoomBrushBase, IHorizontalZoomBrushProps } from "./HorizontalZoomBrush";
import { withSVG } from "../../../hoc";

export const HorizontalZoomBrush = withSVG<IHorizontalZoomBrushProps>(HorizontalZoomBrushBase, "horizontal-zoom-brush");

export type { IHorizontalZoomBrushProps };
