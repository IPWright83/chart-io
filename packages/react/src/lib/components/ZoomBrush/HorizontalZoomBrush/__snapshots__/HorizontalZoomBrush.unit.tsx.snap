// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HorizontalZoomBrush renders correctly 1`] = `
<DocumentFragment>
  <svg>
    <g
      fill="none"
      pointer-events="all"
      transform="translate(0, 200)"
    >
      <g
        class="g-plot line"
      >
        <path
          class="line"
          d="M25,25L50,50                      "
          style="fill: none; pointer-events: none; stroke: rgb(153, 193, 220); stroke-width: 2px;"
        />
        <path
          class="line"
          d="M0,0L100,0"
          style="fill: none; pointer-events: none;"
        />
      </g>
      <rect
        class="overlay"
        cursor="crosshair"
        height="100"
        pointer-events="all"
        width="200"
        x="0"
        y="0"
      />
      <rect
        class="selection"
        cursor="move"
        fill="#777"
        fill-opacity="0.3"
        shape-rendering="crispEdges"
        stroke="#fff"
        style="display: none;"
      />
      <rect
        class="handle handle--w"
        cursor="ew-resize"
        style="display: none;"
      />
      <rect
        class="handle handle--e"
        cursor="ew-resize"
        style="display: none;"
      />
    </g>
  </svg>
</DocumentFragment>
`;
